oauth2-authorization-server/src/main/java/org/springframework/security/oauth2/server/authorization/token/OAuth2TokenCustomizer.java

```diff
@@ -0,0 +1,37 @@
+/*
+ * Copyright 2020 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.springframework.security.oauth2.server.authorization.token;
+
+import java.util.Map;
+
+/**
+ * Basic interface to allow customization of tokens issued by {@link OAuth2TokenIssuer}.
+ *
+ * @author Alexey Nesterov
+ */
+@FunctionalInterface
+public interface OAuth2TokenCustomizer {
+
+	/**
+	 * Apply token customization before it is issued by the issuer.
+	 *
+	 * @param claims original claims created by the issuer
+	 * @param authorizationGrantContext original authorization grant context
+	 */
+	void customizeClaims(Map<String, Object> claims, OAuth2AuthorizationGrantContext authorizationGrantContext);
```

Comment:
I'm wondering if we can simplify this and remove the need for `JwtOAuth2TokenCustomizer`. Maybe we introduce a `OAuth2TokenBuilder` that contains all the data needed. The interface would then be `void customize (OAuth2TokenBuilder tokenBuilder)`. It would be nice to also remove the need for `OAuth2AuthorizationGrantContext` and instead put `RegisteredClient` and `principalName` in a  `Map` of `attributes` in `OAuth2TokenBuilder`. See what you can come up with based on this proposed design.

Reply:
The main purpose was to abstract out notion of headers, which is very specific to JWT and irrelevant for other token types. My concern is that having `OAuth2TokenBuilder` won't solve that, it still will have to include `headers` property. I'll experiment with it and see how it comes together.

Reply:
There are use cases where the headers will need to be customizable, e.g. applications may want to set the `jku` or `x5u` header instead of using `kid`. We'll need to provide this capability to customize headers.

Maybe this would work -> `OAuth2TokenCustomizer.customize(OAuth2TokenBuilder<T extends AbstractOAuth2Token> builder)`. The `Jwt` specific `OAuth2TokenBuilder` would provide a `headers(Consumer<Map<String, Object>> headersConsumer)` or maybe a `headers(Consumer<JoseHeader.Builder> joseHeaderConsumer)`?

Reply:
Ok, thanks for clarifying on headers customization. So I updated the PR with another iteration, trying your idea of introducing a builder. I think it worked out quite nicely. However instead of binding a customizer to a token type by having `T extends AbstractOAuth2Token` I decided to link it to token identity, e.g. Jwt. So a customizers can be used for any token type (Id, Access) if that's represented by a Jwt token.